[NNNN] New proposal for feature profiles (#337)
llvm/wg-hlsl: proposals/NNNN-feature-profiles.md

---
title: "[NNNN] - Feature Profiles"
params:
  authors:
    - llvm-beanz: Chris Bieneman
  status: Under Consideration
  sponsors:
    - llvm-beanz: Chris Bieneman
---

## Introduction

> This proposal seeks to introduce a new more formalized way of specifying feature
> profiles as a set of features to allow the compiler to utilize when generating
> outputs. This proposal pairs stricter availability checks with explicit opt-in

Reviewer: "Feature Profiles" is a new concept being added here?  The current first sentence feels a bit long and clunky.  Here's a stab at another way of wording it:

```suggestion
This proposal introduces feature profiles: a more formalized way to specify a set of features that the compiler can utilize when generating outputs. This proposal pairs stricter availability checks with explicit opt-in
```

mechanisms for optional features to help avoid the compiler introducing
unexpected dependencies on optional features which may not be widely supported.

## Motivation

Today DXC and Clang both implicitly and aggressively opt in to optional features
which may not be supported by all devices. This places a burden on users to
ensure that they are aware of the potential impacts on deployment for shaders
utilizing features which they may or may not have intended.

This feature addresses this problem by defining clear feature profiles derived
from DirectX Shader Models for DirectX targets and [Vulkan
Profiles](https://github.com/KhronosGroup/Vulkan-Profiles) for Vulkan, and
compiler features to enable opting in to optional features that layer on top of
the specified target profile.

## Proposed solution

The core of this proposal revolves around a new definition for "target profiles"
that the shader compiler will translate to available feature sets.

The target profile flag supported by DXC-compatible drivers as `-T` will be
extended to support a new wider set of profile values. The current profiles will
remain unchanged, which will imply a shader-model feature set and _all_ optional

> features.

Reviewer: Since all optional features are included, I wonder if we should have a way to override and remove features.

I think this is implied below with the `+NativeLowPrecision`, example, suggesting that there's a matching `-NativeLowPrecision` option available.  If that's the case it might be nice to explicitly call it out.

Author: Oh, I should have added an example for that. The existing Clang model is `+` to enable `-` to disable. Will update.
